### DISCUSSION TITLE
How to sync repository template changes?

### DISCUSSION THREAD

**Discussion by solvaholic** [community/community#23528] [New to GitHub]
<p>Hiyo! I maintain a repository template to help me set up new repositories. When I make changes to the template, I would like to sync those changes out to all the repositories I’ve created from that template.</p>
<p>For example if I edit .github/workflows/main.yml in my template, I want to automatically create pull requests to bring those changes into each repository I’ve created from that template.</p>
<p>If there’s nothing built-in for this, I think it could be easy to do with GitHub Actions. The ways I’ve found to trigger such a workflow, though, would require running on a schedule or manually keeping a list of repositories I’ve created from the template.</p>
<p>How do you synchronize template changes out to repositories created from the template?</p>

**Comment by logankilpatrick** [comment, marked as answer]
<p>Hey <a class="mention" href="/solvaholic">@solvaholic</a>, there is no way to auto-update the repo’s you have created from the template. In my opinion, the “on push” trigger makes the most sense as long as you push large changes to the repo at once. </p>

**Comment by tetienne** [comment]
<p>Did you manage to do what you asked for?</p>

**Comment by solvaholic** [comment]
<p>Hi <a class="mention" href="/tetienne">@tetienne</a> :wave: I did!</p>
<p>The first solution I tried was to initiate updates from the template repository itself. This required maintaining a list of repositories to sync to, and storing credentials that would let an Actions workflow push to those repositories. That was a lot to keep up with!</p>
<p>Now I include a sync-from-template scheduled workflow in my template repository. This way when I create a repository from the template it will automatically keep itself up-to-date. And I can delete or disable the workflow if don’t want it to run.</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/solvaholic/template/blob/bde824d928e9fb971895862179e15fb70f9fda42/.github/workflows/sync-from-template.yml" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/solvaholic/template/blob/bde824d928e9fb971895862179e15fb70f9fda42/.github/workflows/sync-from-template.yml" target="_blank" rel="noopener nofollow ugc">solvaholic/template/blob/bde824d928e9fb971895862179e15fb70f9fda42/.github/workflows/sync-from-template.yml</a></h4>
<pre><code class="lang-yml">---
#
# - Run this workflow to pull changes from the template repository.
# 
#   - Clone this repository. Check out a branch
#   - Copy files from the template onto this clone
#   - Push the branch to this repository
#   - Create a pull request in this repository
#
name: Sync changes from template

# TODO:
# - Switch to gh. Check https://github.com/cli/cli/issues/297 for updates

on:
  # Run at 0517 UTC each Friday
  schedule:
    - cron: "17 5 * * 5"

  # Run when this file changes
</code></pre>

  This file has been truncated. <a href="https://github.com/solvaholic/template/blob/bde824d928e9fb971895862179e15fb70f9fda42/.github/workflows/sync-from-template.yml" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by marcospgp** [comment]
<p>Isn’t this a clear use case for forking?</p>

**Comment by solvaholic** [comment]
<p>Hi <a class="mention" href="/marcospgp">@marcospgp</a>, thank you for asking about this.</p>
<blockquote>
<p>Isn’t this a clear use case for forking?</p>
</blockquote>
<p>When changes are made to a fork’s parent or root repository, they’re not automatically pushed out to forks. So I think forking wouldn’t automatically sync template changes to repositories forked from the template.</p>
<p>If you see a way for forking to help here, or if I’ve misunderstood what you’re suggesting, please say more :bow:</p>

**Comment by lotyp** [comment]
Found github action for this purpose. Maybe will be helpful, for someone, who googled and reads this 🥹
https://github.com/AndreasAugustin/actions-template-sync

**Reply by nezlicodes** [reply]
thank you buddy  🥹

**Comment by mfhepp** [comment]
An alternative way is to add the template repository as a [Git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules); this allows tracking changes between the updated template and your own code.

For templates that are mainly capturing workflows, Github settings, and other best practices, this may be too much manual effort.
But for templates of e.g. academic research projects or if you typically customize files from the template, it is IMO a worthwhile, safer approach - I would not want an automated merge or similar applied to my derived works.
